assets/js/atomic/blocks/product-elements/button/frontend.tsx

```diff
@@ -0,0 +1,151 @@
+/**
+ * External dependencies
+ */
+import { CART_STORE_KEY as storeKey } from '@woocommerce/block-data';
+import { store } from '@woocommerce/interactivity';
+import { dispatch } from '@wordpress/data';
+import { Cart } from '@woocommerce/type-defs/cart';
+
+type Context = {
+	woocommerce: {
+		isLoading: boolean;
+		addToCartText: string;
+		productId: number;
+		numberOfItems: number;
+	};
+};
+
+type State = {
+	woocommerce: {
+		cart: Cart;
+		inTheCartText: string;
+	};
+};
+
+const getProductById = ( cartState: Cart, productId: number ) => {
+	return cartState.items.find( ( item ) => item.id === productId );
+};
+
+const productButtonSelectors = {
+	woocommerce: {
+		addToCartText: ( {
+			context,
+			state,
+		}: {
+			context: Context;
+			state: State;
+		} ) => {
+			if ( context.woocommerce.numberOfItems === 0 ) {
+				return context.woocommerce.addToCartText;
+			}
+
+			if ( ! state.woocommerce.cart ) {
+				return state.woocommerce.inTheCartText.replace(
+					'###',
+					context.woocommerce.numberOfItems.toString()
+				);
+			}
```

Comment:
Why do you need to maintain two separate implementations? Isn't the point of syncing with Redux to delegate everything to Redux, including the number of items?

Reply:
The redux store isn't initiated. It is initialized only when the Mini Cart block is opened or Checkout/Cart blocks are rendered

Reply:
Doesn't it make sense to initialize it and use it as the source of truth for all the cart state?

Reply:
The thing is that currently, the store is initialized by fetching the Store API. This could cause a flickering effect (the counter of products appears after).

The best approach is that the store redux should be initialized using the data passed via PHP.

Reply:
The flickering can be solved by populating `state.woocommerce.cart` in PHP with the correct shape and information, but I think you should figure out the [caching strategy](https://github.com/woocommerce/woocommerce-blocks/discussions/10153) before attempting to do that.